### PR TITLE
Fix color validation by avoiding null vectors between identical colors & running all checks concurrently

### PR DESCRIPTION
**Pull request by robpvn**
This is a fix for issue #16837 , which also has some (possibly) unintended consequences worth discussing.

If the picked color and a forbidden color are identical (like if they both picked the same palette color and in the special case when a picked color is outside of the allowed range and the method
returns the picked color as the forbidden color), the vector between them is zero and the maths for adjusting the color fails by hitting the iteration limit. This changes the zero vector to the smallest possible vector in order to avoid the issue.

I also added an extra message if you hit the iteration failsafe that picks a random colour. 

However, this can result in some seriously close adjustments in the case of picking identical palette colors, which might be undesirable compared to picking a new palette color. (See picture of adjusted identical palette picks below.) While technically the current behaviour of picking a random new palette color is a bug, I have a feeling many people might see it as a feature. Comments?

### CODE REVIEW

**Comment by robpvn** [comment]
Woops, this actually only fixes the first part of issue #16837, where the range adjustment didn't get done. The second part, where the two checks aren't run concurrently, isn't fixed yet. I can look at that after I get feedback on the identical palette pick question.

**Comment by robpvn** [comment]
Now I've made a fix for the whole issue #16837, which ensures that all checks are made at the same time.

**Comment by abcdefg30** [comment]
Sorry for looking at this only so late. The code changes look good to me on a first glance.

I'd agree that picking a random new colour is worse than just adjusting the colour slightly, so the behaviour in this PR should be fine.

One other thing I noticed: You should remove the `allForbidden` variable now that it is unused.

**Comment by robpvn** [comment]
Don't worry about responding late, it turns out I'm slow to respond as well! :D 

I deleted the unused variable and rebased the commits to the newest bleed.

**Review by abcdefg30 — DISMISSED** [review]
Code changes look good to me and I didn't spot any regressions.

**Comment by abcdefg30** [comment]
I opened #17582 so we don't forget about it once this PR is merged.

File: OpenRA.Mods.Common/ColorValidator.cs
Comment (pchote) [inline OpenRA.Mods.Common/ColorValidator.cs]: Two style issues here:
* Prefer `var` over explicit type names
* Variable name no longer matches usage

IMO the best fix for both issues would be to change `IsValid` to include null check before adding to the HashSet and give a default argument value of null. We can then remove the dummy set here completely.

File: OpenRA.Mods.Common/ColorValidator.cs
Comment (pchote) [inline OpenRA.Mods.Common/ColorValidator.cs]: Likewise here.

File: OpenRA.Mods.Common/ColorValidator.cs
Comment (pchote) [inline OpenRA.Mods.Common/ColorValidator.cs]: `var errorMessages = new HashSet<string>();`

**Comment by pchote** [comment]
LGTM otherwise, and it makes sense to get this in for the playtest so adding to the milestone.

Given the time constraints on the playtest release, I'll apply the fixups and merge this next weekend if not fixed sooner.

**Review by pchote — APPROVED** [review]
Pushed fixups, LGTM now :+1:

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/eea5648b93b6fec926c1d229c9420ffefaa32a43

**Comment by robpvn** [comment]
Sorry about not responding promptly, I was laid up with the flu last week. Thanks for the merge!

**Pull request merged**
